node/pkg/watchers/evm/connectors/batch_poller.go

```diff
@@ -0,0 +1,285 @@
+package connectors
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	"github.com/certusone/wormhole/node/pkg/supervisor"
+	ethEvent "github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/rpc"
+
+	ethereum "github.com/ethereum/go-ethereum"
+
+	"go.uber.org/zap"
+)
+
+// BatchPollConnector uses batch requests to poll for latest, safe and finalized blocks.
+type BatchPollConnector struct {
+	Connector
+	Delay     time.Duration
+	blockFeed ethEvent.Feed
+	errFeed   ethEvent.Feed
+	batchData []BatchEntry
+}
+
+type (
+	Blocks []*NewBlock
+
+	BatchEntry struct {
+		tag      string
+		finality FinalityLevel
+	}
+
+	BatchResult struct {
+		result BlockMarshaller
+		err    error
+	}
+)
+
+const MAX_GAP_BATCH_SIZE uint64 = 5
+
+func NewBatchPollConnector(ctx context.Context, baseConnector Connector, delay time.Duration) (*BatchPollConnector, error) {
+	// Create the batch data in the order we want to report them to the watcher, so finalized is most important, latest is least.
+	batchData := []BatchEntry{
+		{tag: "finalized", finality: Finalized},
+		{tag: "safe", finality: Safe},
```

Comment:
This looks like it will always query for `safe` though the spot where it is used has a boolean flag for if safe is supported. Does something more flexible need to be implemented?